caffe2/mkl/operators/fully_connected_op.cc

```diff
@@ -0,0 +1,108 @@
+#include "caffe2/core/context.h"
+#include "caffe2/core/operator.h"
+#include "caffe2/utils/mkl_utils.h"
+
+#ifdef CAFFE2_HAS_MKL_DNN
+
+namespace caffe2 {
+namespace mkl {
+
+template <typename T>
+class MKLFullyConnectedOp final : public MKLOperator<T> {
+ public:
+
+  MKLFullyConnectedOp(const OperatorDef& operator_def, Workspace* ws)
+      : MKLOperator<T>(operator_def, ws),
+        axis_(OperatorBase::GetSingleArgument<int32_t>("axis", 1)) {}
+  ~MKLFullyConnectedOp() {}
+
+  bool RunOnDevice() override {
+    auto& X = OperatorBase::Input<MKLMemory<float>>(INPUT);
+    auto& filter = OperatorBase::Input<MKLMemory<float>>(FILTER);
+    auto& bias = OperatorBase::Input<MKLMemory<float>>(BIAS);
+    MKLMemory<float>* Y = OperatorBase::Output<MKLMemory<float>>(0);
+        
+    CAFFE_ENFORCE(filter.ndim() == 2, filter.ndim());
+    CAFFE_ENFORCE(bias.ndim() == 1, bias.ndim());
+
+    if (cached_input_dims_ != X.dims() ||
+        cached_filter_dims_ != filter.dims()) {
+      cached_input_dims_ = X.dims();
+      cached_filter_dims_ = filter.dims();
+
+        const int N = filter.dim32(0);
+        CAFFE_ENFORCE(N == bias.dim32(0));
+        
+        auto Y_shape = X.dims();
+        Y_shape[1] = N;        
+        Y_shape.resize(2);
+        
+        size_t inputSizes[4] = {X.dim32(3), X.dim32(2),
+                                X.dim32(1), X.dim32(0)};
+        if(X.ndim()==2)
+        {
+            inputSizes[0] = X.dim32(1);
+            inputSizes[1] = X.dim32(0);          
+        }
+       
+        size_t outputSizes[2] = {Y_shape[1], Y_shape[0]};
+        
+        primitive_.Reset(
+          dnnInnerProductCreateForwardBias<float>,
+          nullptr,
+          X.ndim(),
+          inputSizes,
+        outputSizes[0]);
+
+      Y->Reset(Y_shape, primitive_, dnnResourceDst);
```

Comment:
its one of MKL Resource type. some of them are(dnnResourceSrc, dnnResourceFilter etc) based on which the MKL memory will be created and used. Hope it makes sense.